src/passes/OptimizeInstructions.cpp

```diff
@@ -3543,6 +3543,55 @@ struct OptimizeInstructions
         return inner;
       }
     }
+    {
+      // TODO: Add cancelation for some large constants when shrinkLevel > 0
+      // in FinalOptimizer.
+
+      // (x >> C)  << C   =>   x & -(1 << C)
+      // (x >>> C) << C   =>   x & -(1 << C)
+      Binary* inner;
+      Const *c1, *c2;
+      if (matches(curr,
+                  binary(Shl, binary(&inner, any(), ival(&c1)), ival(&c2))) &&
+          (inner->op == getBinary(inner->type, ShrS) ||
+           inner->op == getBinary(inner->type, ShrU)) &&
+          Bits::getEffectiveShifts(c1) == Bits::getEffectiveShifts(c2)) {
+        auto type = c1->type;
+        if (type == Type::i32) {
+          c1->value = Literal::makeFromInt32(
+            -(1U << Bits::getEffectiveShifts(c1)), Type::i32);
+        } else {
+          c1->value = Literal::makeFromInt64(
+            -(1ULL << Bits::getEffectiveShifts(c1)), Type::i64);
+        }
+        inner->op = getBinary(type, And);
+        return inner;
+      }
+    }
+    {
+      // TODO: Add cancelation for some large constants when shrinkLevel > 0
+      // in FinalOptimizer.
+
+      // (x << C) >>> C   =>   x & (-1 >>> C)
+      // (x >>> C) << C   =>   skip
+      Binary* inner;
+      Const *c1, *c2;
+      if (matches(
+            curr,
+            binary(ShrU, binary(&inner, Shl, any(), ival(&c1)), ival(&c2))) &&
+          Bits::getEffectiveShifts(c1) == Bits::getEffectiveShifts(c2)) {
+        auto type = c1->type;
+        if (type == Type::i32) {
+          c1->value = Literal::makeFromInt32(
+            -1U >> Bits::getEffectiveShifts(c1), Type::i32);
+        } else {
+          c1->value = Literal::makeFromInt64(
+            -1ULL >> Bits::getEffectiveShifts(c1), Type::i64);
```

Comment:
Same as above, we don't need two if arms here I think. (In this case the i64 value would be larger than 32 bits, but it will get truncated automatically.)

Reply:
It looks like this wrong simplification. It reduce:
```wat
(i32.eqz
  (i32.and
    (local.get $0)
    (i32.const 255)
  )
)
```
to
```wat
(local.get $0)
```

Reply:
Can you find out why? It sounds like either you are hitting a bug in the Literal code, or a bug in the change. I would just add a printf to print out the `c1->value` after the change, to see what it contains. Then we should be able to figure out the bug easily.

Reply:
I think it's expected. if we will use unify solution. Let's check how it will work on this example:
```
i32(x) << 24 >>> 24
```
`i32(-1ULL >>> 24)`  ->   `-1` (0xFFFFFFFF)
is not the same as:

`i32(-1U >>> 24)`   ->   `255` (0xFF)

Reply:
So correct unify version should be:
```c++
uint32 driftShift = type == Type::i64 ? 32 : 0;
c1->value = Literal::makeFromInt64(
            -1ULL >> (Bits::getEffectiveShifts(c1) + driftShift), type);
```
And I'm not sure what is clearer. I prefer existing approach

Reply:
I see now, thanks. Yes, the current approach seems better.